src/main/java/fr/xephi/authme/listener/PlayerListener.java

```diff
@@ -103,103 +98,104 @@
-            || permissionsManager.hasPermission(player, PlayerStatePermission.ALLOW_CHAT_BEFORE_LOGIN);
-        if (mayPlayerSendChat) {
-            removeUnauthorizedRecipients(event);
-        } else {
-            event.setCancelled(true);
-            messages.send(player, MessageKey.DENIED_CHAT);
+        // Non-blocking checks
+        try {
+            onJoinVerifier.checkSingleSession(name);
+            onJoinVerifier.checkIsValidName(name);
+        } catch (FailedVerificationException e) {
+            event.setKickMessage(messages.retrieveSingle(name, e.getReason(), e.getArgs()));
+            event.setLoginResult(AsyncPlayerPreLoginEvent.Result.KICK_OTHER);
         }
     }
 
-    private void removeUnauthorizedRecipients(AsyncPlayerChatEvent event) {
-        if (settings.getProperty(RestrictionSettings.HIDE_CHAT)) {
-            event.getRecipients().removeIf(listenerService::shouldCancelEvent);
-            if (event.getRecipients().isEmpty()) {
-                event.setCancelled(true);
-            }
-        }
-    }
+    /*
+     * Login/join/leave events
+     */
 
-    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
-    public void onPlayerMove(PlayerMoveEvent event) {
-        if (settings.getProperty(ALLOW_UNAUTHED_MOVEMENT) && settings.getProperty(ALLOWED_MOVEMENT_RADIUS) <= 0) {
-            return;
-        }
+    // Note: at this stage (HIGHEST priority) the user's permission data should already have been loaded by
+    // the permission handler, we don't need to call permissionsManager.loadUserData()
 
-        /*
-         * Limit player X and Z movements to 1 block
-         * Deny player Y+ movements (allows falling)
-         */
-        Location from = event.getFrom();
-        Location to = event.getTo();
-        if (from.getBlockX() == to.getBlockX()
-            && from.getBlockZ() == to.getBlockZ()
-            && from.getY() - to.getY() >= 0) {
+    @EventHandler(priority = EventPriority.HIGHEST)
+    public void onAsyncPlayerPreLoginEventHighest(AsyncPlayerPreLoginEvent event) {
+        if (event.getLoginResult() != AsyncPlayerPreLoginEvent.Result.ALLOWED) {
             return;
         }
+        final String name = event.getName();
 
-        Player player = event.getPlayer();
-        if (!listenerService.shouldCancelEvent(player)) {
+        if (validationService.isUnrestricted(name)) {
             return;
         }
 
-        if (!settings.getProperty(RestrictionSettings.ALLOW_UNAUTHED_MOVEMENT)) {
-            // "cancel" the event
-            event.setTo(event.getFrom());
-            return;
+        // Slow, blocking checks
+        try {
+            final PlayerAuth auth = dataSource.getAuth(name);
+            final boolean isAuthAvailable = auth != null;
+            onJoinVerifier.checkKickNonRegistered(isAuthAvailable);
+            final JoiningPlayer joiningPlayer = JoiningPlayer.fromName(name);
```

Comment:
Now that `JoiningPlayer#fromPlayer` is no longer used (except in a test) you can drop the class and all methods using it and just use the player name directly to look up permissions with.